services/src/main/java/org/keycloak/organization/authentication/authenticators/browser/OrganizationAuthenticator.java

```diff
@@ -244,18 +246,25 @@ private boolean redirect(AuthenticationFlowContext context, OrganizationModel or
             return false;
         }
 
-        List<IdentityProviderModel> brokers = organization.getIdentityProviders().toList();
+        // first look for an IDP that matches exactly the specified domain
+        IdentityProviderModel idp = organization.getIdentityProviders()
+                .filter(broker -> IdentityProviderRedirectMode.EMAIL_MATCH.isSet(broker) &&
+                    domain.equals(broker.getConfig().get(OrganizationModel.ORGANIZATION_DOMAIN_ATTRIBUTE))).findFirst().orElse(null);
 
-        for (IdentityProviderModel broker : brokers) {
-            if (IdentityProviderRedirectMode.EMAIL_MATCH.isSet(broker)) {
-                String idpDomain = broker.getConfig().get(OrganizationModel.ORGANIZATION_DOMAIN_ATTRIBUTE);
+        if (idp != null) {
+            // redirect the user using the broker that matches the specified domain
+            redirect(context, idp.getAlias(), username);
+            return true;
+        }
 
-                if (domain.equals(idpDomain)) {
-                    // redirect the user using the broker that matches the email domain
-                    redirect(context, broker.getAlias(), username);
-                    return true;
-                }
-            }
+        // look for an idp that can match any of the org domains
+        idp = organization.getIdentityProviders().filter(broker -> IdentityProviderRedirectMode.EMAIL_MATCH.isSet(broker) &&
+                ANY_DOMAIN.equalsIgnoreCase(broker.getConfig().get(OrganizationModel.ORGANIZATION_DOMAIN_ATTRIBUTE)) &&
+                organization.getDomains().map(OrganizationDomainModel::getName).anyMatch(domain::equals)).findFirst().orElse(null);
```

Comment:
nit:
Sorry, to be nitpicky here. This whole expresion is a bit too difficult to understand. Mostly because of the formatting rules I think, but it is just hard to follow. Could you either change this back to a for loop, or use multiple `.filter` instructions to break this up a bit?

Reply:
Sure can!